incubator/orm/orm.go

```diff
@@ -0,0 +1,91 @@
+package orm
+
+import (
+	"io"
+	"reflect"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+const ormCodespace = "orm"
+
+var (
+	// todo: ormCodespace ok or do we need to claim error codes somehow?
+	ErrNotFound         = errors.Register(ormCodespace, 100, "not found")
+	ErrIteratorDone     = errors.Register(ormCodespace, 101, "iterator done")
+	ErrIteratorInvalid  = errors.Register(ormCodespace, 102, "iterator invalid")
+	ErrType             = errors.Register(ormCodespace, 110, "invalid type")
+	ErrUniqueConstraint = errors.Register(ormCodespace, 111, "unique constraint violation")
+	ErrArgument         = errors.Register(ormCodespace, 112, "invalid argument")
+)
+
+type HasKVStore interface {
+	KVStore(key sdk.StoreKey) sdk.KVStore
+}
+
+// Index allows efficient prefix scans is stored as key = concat(indexKeyBytes, rowIDUint64) with value empty
+// so that the row NaturalKey is allows a fixed with 8 byte integer. This allows the MultiKeyIndex key bytes to be
+// variable length and scanned iteratively. The
+type Index interface {
+	Has(ctx HasKVStore, key []byte) bool
+	Get(ctx HasKVStore, key []byte) (Iterator, error)
+	PrefixScan(ctx HasKVStore, start []byte, end []byte) (Iterator, error)
+	ReversePrefixScan(ctx HasKVStore, start []byte, end []byte) (Iterator, error)
+}
+
+// Iterator allows iteration through a sequence of key value pairs
+type Iterator interface {
+	// LoadNext loads the next value in the sequence into the pointer passed as dest and returns the key. If there
+	// are no more items an error is returned
```

Comment:
Agree that the behaviour is not clear without `ErrIteratorDone` mentioned here. Good point. I am also open to drop the error return value and go with panic. The most likely case is probably somebody passing the wrong type to the function. 
What I would like to avoid is that the caller has to check both return values. It should be either the error value to abort or the empty key. 
Until we get some feedback from @aaronc I will just update the doc.